### ISSUE TITLE
Bug or not? Offset on range vector selector does not actually move sample timestamps

### ISSUE THREAD

**Issue by juliusv** [prometheus/prometheus#12346]
### What did you do?

(Since I think this was never something we talked explicitly about, I'm not sure if I'm reporting a bug or just have always had different expectations about the undefined / undocumented behavior of offsets on range vector selectors.)

I evaluated a range vector selector with an offset of 1h, e.g.:

https://demo.promlabs.com/graph?g0.expr=demo_memory_usage_bytes%7Btype%3D%22free%22%2Cinstance%3D%22demo-service-0%3A10000%22%7D%5B1m%5D%20offset%201h&g0.tab=1&g0.stacked=0&g0.show_exemplars=0&g0.range_input=1h&g0.end_input=2023-05-09%2015%3A56%3A32&g0.moment_input=2023-05-09%2015%3A56%3A32

### What did you expect to see?

I expected the offsetted selector to not only select a 1m range from 1h ago, but to also move the timestamps of the selected data points by 1h into the future (so it looks like they just happened, similar to how an offset on an instant vector selector behaves).

### What did you see instead? Under which circumstances?

The sample timestamps are still from 1h ago. I think I never noticed this because in practice it rarely matters - almost always, the range vector is immediately fed into a function like `rate()` that then emits a single instant vector at the evaluation timestamp, so the behavior becomes invisible. But is this behavior really the expected one?

**Comment by juliusv** [comment]
To be fair, when using the `timestamp()` function (which has hardcoded special-case behavior to avoid having its input timestamp set to the evaluation timestamp) on the result of an offsetted instant vector selector, it behaves the same and does not move the original sample timestamp to the current time. Is this just how it's supposed to work?

**Comment by beorn7** [comment]
I think everything is working as expected.

A vector result type comes back with the evaluation timestamp (which is, in a way, the reason why you need a `timestamp` function at all). A range query creates a matrix result type with the regularly spaced evaluation timestamps. You cannot do a range query with a range selector as the final result, though. But you can query a range selector in an instant query, yielding a matrix result type with timestamps from the actual TSDB. This provides a "trick" to display actual timestamps from the TSDB in the Prometheus UI. I don't think `offset` should manipulate the timestamps returned in that way. They simply never were meant to be evaluation timestamps.

**Comment by juliusv** [comment]
Yeah, reading https://prometheus.io/docs/prometheus/latest/querying/basics/#offset-modifier I guess it also really doesn't mention anything about time-shifting the output timestamps of offsetted range vectors, but our docs are woefully lacking in specific details anyway (e.g. there's nothing in there about the 5-minute lookback delta behavior either).

If this is the expected behavior, then all good, I just lived with the wrong assumption about what the offsetting should "reasonably" do to the underlying sample timestamps for many years, and for instant vector results it even looks like that as a side effect of what instant vector selector results do anyway to their selected output sample :)

Closing for now.
